### PR TITLE
FOCI feature and AppMetadata support in cache

### PR DESCRIPTION
**Pull request by sameerag**
This PR has changes to:
* Create `AppMetadata` cache when `Foci` is sent in the sever response
* Modified `CacheManager` to read cache data with helpers
* Added unit tests

### CODE REVIEW

File: lib/msal-common/src/cache/CacheManager.ts [hunk 1]
```diff
@@ -493,68 +421,233 @@ export abstract class CacheManager implements ICacheManager {
...
+        return refreshTokens[0] as RefreshTokenEntity;
+    }
+
+    /**
+     * Helper to retrieve the appropriate refresh token from cache
+     * @param clientId
+     * @param account
+     * @param familyRT
+     */
+    readRefreshTokenFromCacheHelper(clientId: string, account: AccountInfo, familyRT: boolean): RefreshTokenEntity | null {
```
Comment (tnorling) [inline lib/msal-common/src/cache/CacheManager.ts:551]: This looks the same as `readRefreshTokenFromCache` above?

File: lib/msal-common/src/client/RefreshTokenClient.ts [hunk 2]
```diff
@@ -62,20 +63,64 @@ export class RefreshTokenClient extends BaseClient {
...
             throw ClientAuthError.createNoAccountInSilentRequestError();
-        } 
+        }
 
-        const refreshToken = this.cacheManager.getRefreshTokenEntity(this.config.authOptions.clientId, request.account);
-        // no refresh Token
-        if (!refreshToken) {
-            throw ClientAuthError.createNoTokensFoundError();
+        // try checking if Family Refresh Token (FRT) is available
+        const familyRTResult = await this.acquireFamilyRefreshToken(request);
```
Comment (tnorling) [inline lib/msal-common/src/client/RefreshTokenClient.ts:69]: I would suggest having `acquireFamilyRefreshToken` return the refresh token and then move line 77 down to outside the if/else

File: lib/msal-common/src/cache/CacheManager.ts (hunk 1 above)
Comment (sameerag) [reply lib/msal-common/src/cache/CacheManager.ts:551]: This is not longer needed, let me update. Thanks for the catch

**Review by tnorling — APPROVED** [review]
Looks good! Thanks for the tests

File: lib/msal-common/src/client/RefreshTokenClient.ts (hunk 2 above)
Comment (sameerag) [reply lib/msal-common/src/client/RefreshTokenClient.ts]: Refactored

File: change/@azure-msal-common-2020-09-21-10-44-34-appmetadata-foci.json [hunk 3]
```diff
@@ -0,0 +1,8 @@
+{
+  "type": "minor",
+  "comment": "FOCI - Family of Client IDs feature",
```
Comment (jo-arroyo) [inline change/@azure-msal-common-2020-09-21-10-44-34-appmetadata-foci.json:3]: Nit: Are we putting pull request numbers in the comments?

File: lib/msal-common/src/client/RefreshTokenClient.ts [hunk 4]
```diff
@@ -78,6 +107,11 @@ export class RefreshTokenClient extends BaseClient {
         return this.acquireToken(refreshTokenRequest);
     }
 
+    /**
+     *
```
Comment (jo-arroyo) [inline lib/msal-common/src/client/RefreshTokenClient.ts:111]: Was there going to be a comment here?

**Review by jo-arroyo — APPROVED** [review]
LGTM, with a few questions

File: lib/msal-common/src/client/RefreshTokenClient.ts [hunk 5]
```diff
@@ -62,9 +61,39 @@ export class RefreshTokenClient extends BaseClient {
...
+            catch (e) {
+                // if family Refresh Token (FRT) cache acquisition fails, reattempt with application Refresh Token (ART)
+                if (e instanceof ClientAuthError && e.errorCode === ClientAuthErrorMessage.noTokensFoundError.code) {
+                    return this.acquireTokenWithCachedRefreshToken(request, false);
+                }
+                // if client_mismatch, retry with ART
+                if (!(e instanceof ServerError && e.errorCode === Errors.INVALID_GRANT_ERROR && e.subError === Errors.CLIENT_MISMATCH_ERROR)) {
+                    throw e;
+                }
+            }
```
Comment (sangonzal) [inline lib/msal-common/src/client/RefreshTokenClient.ts:83]:
```suggestion
                const fociRtNotFound = e instanceof ClientAuthError && e.errorCode === ClientAuthErrorMessage.noTokensFoundError.code;
                const clientMismatchError = e instanceof ServerError && e.errorCode === Errors.INVALID_GRANT_ERROR && e.subError === Errors.CLIENT_MISMATCH_ERROR;
                if (rtNotFound || clientMismatchError) {
                    return this.acquireTokenWithCachedRefreshToken(request, false);
                } 
                    throw e;
            }
```

File: lib/msal-common/src/client/RefreshTokenClient.ts [hunk 6]
```diff
@@ -78,6 +107,11 @@ export class RefreshTokenClient extends BaseClient {
         return this.acquireToken(refreshTokenRequest);
     }
 
+    /**
+     *
+     * @param request
+     * @param authority
+     */
```
Comment (sangonzal) [inline lib/msal-common/src/client/RefreshTokenClient.ts:114]: nit: Adds no additional information, and instead clutters the code. If you add comments to methods, consider filling them in with info.

File: lib/msal-common/src/utils/Constants.ts [hunk 7]
```diff
@@ -271,6 +272,7 @@ export enum CacheType {
  */
 export const APP_METADATA = "appmetadata";
 export const ClientInfo = "client_info";
+export const THE_FAMILY_ID = "1";
```
Comment (sangonzal) [inline lib/msal-common/src/utils/Constants.ts:275]: nit: wondering if theres a reason why it's THE_FAMILY_ID, as opposed to just FAMILIY_ID?

**Review by sangonzal — APPROVED** [review]

File: lib/msal-common/src/client/RefreshTokenClient.ts (hunk 6 above)
Comment (sameerag) [reply lib/msal-common/src/client/RefreshTokenClient.ts:114]: This is a miss. Will add now. thanks for catching it

File: lib/msal-common/src/utils/Constants.ts (hunk 7 above)
Comment (sameerag) [reply lib/msal-common/src/utils/Constants.ts:275]: As we have only one family id. :-D and nothing is going to change unless we break a ton of apps out there. As far as I know, they want to get out of FOCI model than add more families.

File: lib/msal-common/src/client/RefreshTokenClient.ts (hunk 4 above)
Comment (sameerag) [reply lib/msal-common/src/client/RefreshTokenClient.ts:111]: yes. Missed adding, will add now. Thanks for the catch.

File: change/@azure-msal-common-2020-09-21-10-44-34-appmetadata-foci.json (hunk 3 above)
Comment (sameerag) [reply change/@azure-msal-common-2020-09-21-10-44-34-appmetadata-foci.json:3]: FOCI is a feature name. Do you mean we should add the PR number? Am I missing something?

File: lib/msal-common/src/cache/CacheManager.ts [hunk 8]
```diff
@@ -174,12 +148,29 @@ export abstract class CacheManager implements ICacheManager {
...
+        // don't parse any non-account type cache entities
+        if (CredentialEntity.getCredentialType(key) !== Constants.NOT_DEFINED || this.isAppMetadata(key)) {
+            return null;
+        }
+
+        // Attempt retrieval
+        let entity: AccountEntity;
+        try {
+            entity = this.getItem(key, CacheSchemaType.ACCOUNT) as AccountEntity;
+        } catch (e) {
```
[the remark this answers is not in the document]
Comment (sameerag) [reply lib/msal-common/src/cache/CacheManager.ts:164]: Currently logger isn't accessible in CacheManager. Will add the instrumentation when the pipeline gets added for logging in all cache entities. Also ... on second thought, there may not be an exception anymore for `getItem()`, it should simply return null if an account is not found and I believe we handle that. Will check again and confirm.

File: lib/msal-common/src/client/RefreshTokenClient.ts (hunk 5 above)
Comment (sameerag) [reply lib/msal-common/src/client/RefreshTokenClient.ts:83]: Addressed. Thanks.

File: lib/msal-common/src/client/RefreshTokenClient.ts (hunk 6 above)
Comment (sameerag) [reply lib/msal-common/src/client/RefreshTokenClient.ts:114]: Done

File: lib/msal-common/src/client/RefreshTokenClient.ts (hunk 4 above)
Comment (sameerag) [reply lib/msal-common/src/client/RefreshTokenClient.ts:111]: Done

File: lib/msal-common/src/cache/CacheManager.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (sameerag) [reply lib/msal-common/src/cache/CacheManager.ts:164]: There is an upcoming PR that changes `ICacheManager` to improve browser storage API. Will address these concerns on exceptions (or rather remove them) as a part of that PR. Closing this for now.

**Pull request merged**